File: Sources/DiskMapCore/CirclePack.swift
import Foundation

public struct PackedCircle: Sendable, Equatable {
    public var id: String
    public var x: Double
    public var y: Double
    public var radius: Double

    public init(id: String, x: Double, y: Double, radius: Double) {
        self.id = id
        self.x = x
        self.y = y
        self.radius = radius
    }
}

/// Circle packing for the bubbles view. Radii come from `sqrt(size)` so
/// area tracks size. Each sibling is seated tangent to an already placed
/// circle, at the angle closest to the origin that clears the others. A
/// separation pass then pushes any remaining intersection apart, so a
/// missed sample cannot leave two circles overlapping. The enclosing
/// circle may contain empty space.
public enum CirclePack {
    public static func pack(_ slices: [ChartSlice]) -> [PackedCircle] {
        packNodes(slices).circles
    }

    private struct Node {
        var id: String
        var radius: Double
        var x: Double
        var y: Double
        var children: [Node]
    }

    private static func packNodes(_ slices: [ChartSlice]) -> (circles: [PackedCircle], enclosing: Double) {
        let nodes = slices.filter { $0.size > 0 }.sorted { $0.size > $1.size }.map { slice -> Node in
            if slice.children.isEmpty {
                return Node(id: slice.id, radius: sqrt(Double(slice.size)), x: 0, y: 0, children: [])
            }
            let nested = packNodes(slice.children)
            let radius = max(sqrt(Double(slice.size)), nested.enclosing)
            return Node(id: slice.id, radius: radius, x: 0, y: 0, children: scaled(nested.circles, into: radius))
        }
        var placed = nodes
        placeSiblings(&placed)
        var circles: [PackedCircle] = []
        for node in placed {
            circles.append(PackedCircle(id: node.id, x: node.x, y: node.y, radius: node.radius))
            for child in node.children {
                circles.append(PackedCircle(
                    id: child.id,
                    x: node.x + child.x,
                    y: node.y + child.y,
                    radius: child.radius
                ))
            }
        }
        return (circles, enclosingRadius(placed))
    }

    private static func scaled(_ circles: [PackedCircle], into enclosing: Double) -> [Node] {
        let current = circles.reduce(0) { radius, circle in
            max(radius, hypot(circle.x, circle.y) + circle.radius)
        }
        let scale = current > 0 ? min(1, enclosing / current) : 1
        return circles.map { circle in
            Node(id: circle.id, radius: circle.radius * scale, x: circle.x * scale, y: circle.y * scale, children: [])
        }
    }

    private static func enclosingRadius(_ nodes: [Node]) -> Double {
        nodes.reduce(0) { radius, node in
            max(radius, hypot(node.x, node.y) + node.radius)
        }
    }

    private static func placeSiblings(_ nodes: inout [Node]) {
        guard !nodes.isEmpty else { return }
        nodes[0].x = 0
        nodes[0].y = 0
        guard nodes.count > 1 else { return }
        nodes[1].x = nodes[0].radius + nodes[1].radius
        nodes[1].y = 0
        let samples = 72
        for index in 2..<nodes.count {
            var bestX = nodes[index - 1].x + nodes[index - 1].radius + nodes[index].radius
            var bestY = 0.0
            var bestScore = Double.greatestFiniteMagnitude
            for host in 0..<index {
                let distance = nodes[host].radius + nodes[index].radius
                for step in 0..<samples {
                    let angle = Double(step) / Double(samples) * 2 * .pi
                    let x = nodes[host].x + cos(angle) * distance
                    let y = nodes[host].y + sin(angle) * distance
                    guard clears(x: x, y: y, radius: nodes[index].radius, nodes: nodes, count: index) else { continue }
                    let score = x * x + y * y
                    if score < bestScore {
                        bestScore = score
                        bestX = x
                        bestY = y
                    }
                }
            }
            nodes[index].x = bestX
            nodes[index].y = bestY
        }
        separate(&nodes)
    }

    private static func clears(x: Double, y: Double, radius: Double, nodes: [Node], count: Int) -> Bool {
        for index in 0..<count {
            let dx = x - nodes[index].x
            let dy = y - nodes[index].y
            let minDist = radius + nodes[index].radius
            if dx * dx + dy * dy < minDist * minDist - 1e-6 { return false }
        }
        return true
    }

    private static func separate(_ nodes: inout [Node]) {
        guard nodes.count > 1 else { return }
        for _ in 0..<100 {
            var moved = false
        for i in 0..<nodes.count {
            for j in (i + 1)..<nodes.count {
                    let dx = nodes[j].x - nodes[i].x
                    let dy = nodes[j].y - nodes[i].y
                    let dist = hypot(dx, dy)
                    let minDist = nodes[i].radius + nodes[j].radius
                    guard dist < minDist - 1e-6 else { continue }
                    let push = (minDist - dist) / 2 + 1e-4
                    let ux = dist > 0 ? dx / dist : 1
                    let uy = dist > 0 ? dy / dist : 0
                    nodes[i].x -= ux * push
                    nodes[i].y -= uy * push
                    nodes[j].x += ux * push
                    nodes[j].y += uy * push
                    moved = true
                }
            }
            if !moved { return }
        }
    }
}

File: windows/src/DiskMap.Core/Layout/CirclePack.cs
namespace DiskMap.Core;

public readonly record struct PackedCircle(string Id, double X, double Y, double Radius);

/// <summary>
/// Circle packing for the bubbles view. Radii come from sqrt(size) so
/// area tracks size. Each sibling is seated tangent to an already placed
/// circle, at the angle closest to the origin that clears the others. A
/// separation pass then pushes any remaining intersection apart, so a
/// missed sample cannot leave two circles overlapping. The enclosing
/// circle may contain empty space.
/// </summary>
public static class CirclePack
{
    public static List<PackedCircle> Pack(List<ChartSlice> slices) => PackNodes(slices).Circles;

    private sealed class Node
    {
        public required string Id;
        public required double Radius;
        public double X;
        public double Y;
        public List<Node> Children = [];
    }

    private static (List<PackedCircle> Circles, double Enclosing) PackNodes(List<ChartSlice> slices)
    {
        var nodes = slices
            .Where(s => s.Size > 0)
            .OrderByDescending(s => s.Size)
            .Select(slice =>
            {
                if (slice.Children.Count == 0)
                {
                    return new Node { Id = slice.Id, Radius = Math.Sqrt(slice.Size) };
                }
                var nested = PackNodes(slice.Children);
                double radius = Math.Max(Math.Sqrt(slice.Size), nested.Enclosing);
                return new Node { Id = slice.Id, Radius = radius, Children = ScaleInto(nested.Circles, radius) };
            })
            .ToList();

        PlaceSiblings(nodes);
        var circles = new List<PackedCircle>();
        foreach (var node in nodes)
        {
            circles.Add(new PackedCircle(node.Id, node.X, node.Y, node.Radius));
            foreach (var child in node.Children)
                circles.Add(new PackedCircle(child.Id, node.X + child.X, node.Y + child.Y, child.Radius));
        }
        return (circles, EnclosingRadius(nodes));
    }

    private static List<Node> ScaleInto(List<PackedCircle> circles, double enclosing)
    {
        double current = circles.Aggregate(0.0, (r, c) => Math.Max(r, Math.Sqrt(c.X * c.X + c.Y * c.Y) + c.Radius));
        double scale = current > 0 ? Math.Min(1, enclosing / current) : 1;
        return circles.Select(c => new Node { Id = c.Id, Radius = c.Radius * scale, X = c.X * scale, Y = c.Y * scale }).ToList();
    }

    private static double EnclosingRadius(List<Node> nodes) =>
        nodes.Aggregate(0.0, (r, n) => Math.Max(r, Math.Sqrt(n.X * n.X + n.Y * n.Y) + n.Radius));

    private static void PlaceSiblings(List<Node> nodes)
    {
        if (nodes.Count == 0) return;
        nodes[0].X = 0;
        nodes[0].Y = 0;
        if (nodes.Count == 1) return;
        nodes[1].X = nodes[0].Radius + nodes[1].Radius;
        nodes[1].Y = 0;
        const int samples = 72;
        for (int index = 2; index < nodes.Count; index++)
        {
            double bestX = nodes[index - 1].X + nodes[index - 1].Radius + nodes[index].Radius;
            double bestY = 0;
            double bestScore = double.MaxValue;
            for (int host = 0; host < index; host++)
            {
                double distance = nodes[host].Radius + nodes[index].Radius;
                for (int step = 0; step < samples; step++)
                {
                    double angle = (double)step / samples * 2 * Math.PI;
                    double x = nodes[host].X + Math.Cos(angle) * distance;
                    double y = nodes[host].Y + Math.Sin(angle) * distance;
                    if (!Clears(x, y, nodes[index].Radius, nodes, index)) continue;
                    double score = x * x + y * y;
                    if (score < bestScore)
                    {
                        bestScore = score;
                        bestX = x;
                        bestY = y;
                    }
                }
            }
            nodes[index].X = bestX;
            nodes[index].Y = bestY;
        }
        Separate(nodes);
    }

    private static bool Clears(double x, double y, double radius, List<Node> nodes, int count)
    {
        for (int i = 0; i < count; i++)
        {
            double dx = x - nodes[i].X;
            double dy = y - nodes[i].Y;
            double minDist = radius + nodes[i].Radius;
            if (dx * dx + dy * dy < minDist * minDist - 1e-6) return false;
        }
        return true;
    }

    private static void Separate(List<Node> nodes)
    {
        if (nodes.Count <= 1) return;
        for (int iter = 0; iter < 100; iter++)
        {
            bool moved = false;
            for (int i = 0; i < nodes.Count; i++)
            {
                for (int j = i + 1; j < nodes.Count; j++)
                {
                    double dx = nodes[j].X - nodes[i].X;
                    double dy = nodes[j].Y - nodes[i].Y;
                    double dist = Math.Sqrt(dx * dx + dy * dy);
                    double minDist = nodes[i].Radius + nodes[j].Radius;
                    if (dist >= minDist - 1e-6) continue;
                    double push = (minDist - dist) / 2 + 1e-4;
                    double ux = dist > 0 ? dx / dist : 1;
                    double uy = dist > 0 ? dy / dist : 0;
                    nodes[i].X -= ux * push;
                    nodes[i].Y -= uy * push;
                    nodes[j].X += ux * push;
                    nodes[j].Y += uy * push;
                    moved = true;
                }
            }
            if (!moved) return;
        }
    }
}
